### ISSUE TITLE
EMQX Max Conn Rate Limiter does not work. Hangs the entire listener with 100% CPU utilization

### ISSUE THREAD

**Issue by thehellmaker** [emqx/emqx#14145]
### What happened?

We have an EMQX cluster to which IOT devices connect. Out SSL listener configuration is as below
```
listeners.ssl.default {
    enable_authn = false
    bind = "0.0.0.0:8883"
    acceptors = 64
    max_connections = 1024000
    max_conn_rate = "10/s"
    messages_rate = "10/s"
    bytes_rate = "1MB/s"
    ssl_options {
        ....
    }
}
```

Lets say i have 2000 devices connected to the EMQX cluster and they all work well without any issues until there is an outage are emqx service bounce when all devices get disconnected and reconnect back simultaneously. Even though we have max_conn_rate set to 10/s EMQX is unable to handle this burst of connection requests simultaneously. 

There are a few symptoms we see here
1. When EMQX is bounced and there is a time window before emqx fully boots up. But our devices are aggressive they keep trying to reconnect to the cluster ASAP we see during this time that sometimes 0 devices connect back and sometimes upto 10 devices connect back and rest of the devices are unable to connect back and logs are as shown below
```

2024-10-21T04:53:43.479507+00:00 [warning] msg: alarm_is_activated, mfa: emqx_alarm:do_actions/3(418), message: <<"VM is overloaded on node: 'emqx@127.0.0.1': 141">>, name: runq_overload
2024-10-21T04:54:08.265984+00:00 [warning] msg: alarm_is_deactivated, mfa: emqx_alarm:do_actions/3(424), name: runq_overload

```

2. We run on AWS and when we block the route to this server during boot up so that no devices interrupts the healthy bootup on EMQX and then open the port for the devices to connect then we see between 20 and 50 device connecting at different times but then the other devices are unable to connect to the cluster and we see the same message 

```

2024-10-21T04:53:43.479507+00:00 [warning] msg: alarm_is_activated, mfa: emqx_alarm:do_actions/3(418), message: <<"VM is overloaded on node: 'emqx@127.0.0.1': 141">>, name: runq_overload
2024-10-21T04:54:08.265984+00:00 [warning] msg: alarm_is_deactivated, mfa: emqx_alarm:do_actions/3(424), name: runq_overload

```
During this time we also see that beam.smp CPU utilization is 100% but no devices are able to connect on this listener.. However there are other listeners where devices do not connect but few connections exist and this spike doesn't happen and those listeners still respond well.

### What did you expect to happen?

EMQX gracefully handled rate limiting and allows max_conn_rate devices to connect back to the cluster and pace out the connections. 

### How can we reproduce it (as minimally and precisely as possible)?

Simultaneously establish 1500+ Mutual TLS Auth to the EMQX listener and this issue is easily reproducible. Ofcourse this could be CPU bound and on bigger servers the limit might be larger. One of the benchmarks we ran for testing was r7gd.medium host on AWS adn 1500+ Mutual TLS connections. 

### Anything else we need to know?

This got resolved when we deployed an L4 Proxy with rate limiting using token bucket in front of EMQX and all devices were able to ramp up and successfully able to connecte.

### EMQX version

<details>

```console
$ ./bin/emqx_ctl broker
sysdescr  : EMQX
version   : 5.5.0
datetime  : 2024-11-05T08:53:35.155015788+00:00
uptime    : 22 hours, 55 minutes, 7 seconds
```

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
Linux ip-****-**-aws #56~20.04.1-Ubuntu SMP Tue Nov 28 15:43:06 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
```

</details>

### Log files

<details>

</details>

**Comment by zmstone** [comment]
Hi @thehellmaker 
I updated the subject of this issue,
then realize that you have `bytes_rate = "1MB/s"`, to help narrow down the investigation scope, could you confirm that your tests were only client connect/reconnect, but no messages published ?

if you do have messages published, could you remove the bytes_rate limit and test again?

**Comment by thehellmaker** [comment]
Device do publish sync messages on the topics and subscribe to their topics upon connection. Do you mean if bytes_rate is provided then max_conn_rate is not honoured?

**Comment by zmstone** [comment]
I mean, I am unsure if it's `max_conn_rate` or `byte_rate` causing the high load, it would be great if you can help to find out.

**Comment by thehellmaker** [comment]
Sure just to clarify you want me to change to the below config and verify.
```
listeners.ssl.default {
    enable_authn = false
    bind = "0.0.0.0:8883"
    acceptors = 64
    max_connections = 1024000
    max_conn_rate = "10/s"
    messages_rate = "10/s"
    ssl_options {
        ....
    }
}
```

**Comment by thehellmaker** [comment]
Same Issue with the suggested configuration

```
2024-11-06T07:03:09.242277+00:00 [warning] msg: alarm_is_activated, mfa: emqx_alarm:do_actions/3(418), message: <<"VM is overloaded on node: 'emqx@127.0.0.1': 142">>, name: runq_overload
2024-11-06T07:03:15.592476+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:03:15.633068+00:00 [warning] msg: alarm_is_activated, mfa: emqx_alarm:do_actions/3(418), message: <<"resource down: timeout">>, name: <<"emqx_authn_http:2">>
2024-11-06T07:03:15.686044+00:00 [warning] msg: health_check_failed, mfa: emqx_resource_manager:handle_connected_health_check/1(893), id: <<"emqx_authn_http:2">>, status: disconnected
2024-11-06T07:03:30.717311+00:00 [warning] msg: emqx_connector_on_start_already_started, mfa: emqx_bridge_http_connector:start_pool/2(256), pool_name: <<"emqx_authn_http:2">>
2024-11-06T07:03:30.744003+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:03:31.020903+00:00 [warning] msg: start_resource_failed, mfa: emqx_resource_manager:start_resource/2(612), id: <<"emqx_authn_http:2">>, reason: timeout
2024-11-06T07:03:46.583830+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:03:46.918098+00:00 [warning] msg: start_resource_failed, mfa: emqx_resource_manager:start_resource/2(612), id: <<"emqx_authn_http:2">>, reason: timeout
2024-11-06T07:04:02.671281+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:04:03.334661+00:00 [warning] msg: start_resource_failed, mfa: emqx_resource_manager:start_resource/2(612), id: <<"emqx_authn_http:2">>, reason: timeout
2024-11-06T07:04:18.690546+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:04:18.785919+00:00 [warning] msg: start_resource_failed, mfa: emqx_resource_manager:start_resource/2(612), id: <<"emqx_authn_http:2">>, reason: timeout
2024-11-06T07:04:34.292007+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:04:34.983716+00:00 [warning] msg: alarm_is_activated, mfa: emqx_alarm:do_actions/3(418), message: <<"81.43% cpu usage">>, name: high_cpu_usage
2024-11-06T07:04:49.321384+00:00 [warning] msg: emqx_connector_on_start_already_started, mfa: emqx_bridge_http_connector:start_pool/2(256), pool_name: <<"emqx_authn_http:2">>
2024-11-06T07:04:49.351852+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:04:49.684821+00:00 [warning] msg: start_resource_failed, mfa: emqx_resource_manager:start_resource/2(612), id: <<"emqx_authn_http:2">>, reason: timeout
2024-11-06T07:05:05.430147+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:05:05.827543+00:00 [warning] msg: start_resource_failed, mfa: emqx_resource_manager:start_resource/2(612), id: <<"emqx_authn_http:2">>, reason: timeout
2024-11-06T07:05:22.863985+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:05:23.658681+00:00 [warning] msg: start_resource_failed, mfa: emqx_resource_manager:start_resource/2(612), id: <<"emqx_authn_http:2">>, reason: timeout
2024-11-06T07:05:32.185002+00:00 [warning] msg: alarm_is_deactivated, mfa: emqx_alarm:do_actions/3(424), name: <<"emqx_authn_http:2">>
2024-11-06T07:05:51.923119+00:00 [warning] msg: http_connector_do_request_failed, mfa: emqx_bridge_http_connector:on_query/3(387), peername: 49.206.2.230:65180, clientid: app-AV0Kj3htXZbpcDOzLoV3jNDZTg42-dd6a1789-5006-427c-b80a-9d046fa261f3, reason: timeout, connector: <<"emqx_authn_http:2">>
2024-11-06T07:05:51.923344+00:00 [error] msg: send_error, mfa: emqx_resource_buffer_worker:handle_query_result_pure/3(942), peername: 49.206.2.230:65180, clientid: app-AV0Kj3htXZbpcDOzLoV3jNDZTg42-dd6a1789-5006-427c-b80a-9d046fa261f3, id: <<"emqx_authn_http:2">>, reason: {recoverable_error,timeout}
2024-11-06T07:06:59.542513+00:00 [error] msg: http_connector_pmap_failed, mfa: emqx_bridge_http_connector:do_get_status/2(554), reason: timeout
2024-11-06T07:06:59.561536+00:00 [warning] msg: alarm_is_activated, mfa: emqx_alarm:do_actions/3(418), message: <<"resource down: timeout">>, name: <<"emqx_authn_http:2">>
2024-11-06T07:06:59.573767+00:00 [warning] msg: health_check_failed, mfa: emqx_resource_manager:handle_connected_health_check/1(893), id: <<"emqx_authn_http:2">>, status: disconnected
2024-11-06T07:07:14.591264+00:00 [warning] msg: emqx_connector_on_start_already_started, mfa: emqx_bridge_http_connector:start_pool/2(256), pool_name: <<"emqx_authn_http:2">>
2024-11-06T07:07:14.591721+00:00 [warning] msg: alarm_is_deactivated, mfa: emqx_alarm:do_actions/3(424), name: <<"emqx_authn_http:2">>
2024-11-06T07:07:26.653121+00:00 [warning] msg: http_connector_do_request_failed, mfa: emqx_bridge_http_connector:on_query/3(387), peername: 171.48.96.133:11009, clientid: app-yQH7yMWp0tSLgKUdCOw84Wn5iqR2-aa2d7cdc-e70d-4fdd-88b6-24a4129f0a5b, reason: timeout, connector: <<"emqx_authn_http:2">>
2024-11-06T07:07:26.653281+00:00 [error] msg: send_error, mfa: emqx_resource_buffer_worker:handle_query_result_pure/3(942), peername: 171.48.96.133:11009, clientid: app-yQH7yMWp0tSLgKUdCOw84Wn5iqR2-aa2d7cdc-e70d-4fdd-88b6-24a4129f0a5b, id: <<"emqx_authn_http:2">>, reason: {recoverable_error,timeout}
```

Even thought we dnt provide it in the configuration messages_rate and bytes rate are defaulted in from the listener config from the UI 
<img width="962" alt="Screenshot 2024-11-06 at 12 36 19 PM" src="https://github.com/user-attachments/assets/c199ea49-4211-4831-955f-cb6b8871e72d">

**Comment by qzhuyan** [comment]
Will be helpful if you could confirm if latest version 5.8.1 has the same issue or not.
If so, I could dig into that in 5.8.1 which is my main focus.

**Comment by zmstone** [comment]
@thehellmaker 
Thank you for the update. will work on this soon.

**Comment by zmstone** [comment]
@holopekochan thank you for contributing to the discussions.
I'd like to know if you have noticed high CPU utilization in your case as well?
I am able to reproduce what you have observed, but not a high CPU utilization.

Here is my findings
When the acceptor is suspended (due to rate limit), the connections will stay in the backlog
if stay there for long enough, the client will give up waiting, socket is closed, but EMQX does not know
EMQX will will still try to accept an in-fact half-closed socket and consume the rate limit credit.
This is a known issue and in our TODO list, will make it higher priority.

If you check `netstat -lnpt | grep 1883` the second column is the number of connections in the backlog.
Before the fix, firewall can be used to control the TCP syn rate.
Here is an iptables based example previously verified by @qzhuyan 

```
iptables -N EMQX_TCP
iptables -N EMQX_TCP_SYN
iptables -A INPUT -i veth2 -p tcp -j EMQX_TCP
iptables -A EMQX_TCP -p tcp -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -j EMQX_TCP_SYN
iptables -A EMQX_TCP_SYN -p tcp -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -m limit --limit 20/sec -j ACCEPT
iptables -A EMQX_TCP_SYN -j DROP
```

**Comment by thehellmaker** [comment]
@zmstone I can confirm the high CPU utilization is not observed in 5.8.1. Rest of the issues are valid.

**Comment by zmstone** [comment]
> @zmstone I can confirm the high CPU utilization is not observed in 5.8.1. Rest of the issues are valid.

Thank you. 
I'm working on the fix.

**Comment by thehellmaker** [comment]
Thanks for the quick fix and update. Is my understanding correct that this fix will be a part of 5.8.3?

**Comment by zmstone** [comment]
related: https://github.com/emqx/emqtt-bench/pull/270
